File: src/samples.ts
/** Typed sample banks (`samples Tracks { … }`) — lookup helpers for validate / evaluate / resolve. */

import type { SampleFieldDecl, SampleEntryDecl, SamplesDecl } from "./ast.js";
import type { DesignDefinition } from "./designModel.js";
import { PdlError } from "./errors.js";

/** Split `Bank.entry.field` into three segments. Longer/shorter dotted paths return `undefined`. */
export function splitSamplePath(path: string): [string, string, string] | undefined {
  const parts = path.split(".");
  if (parts.length !== 3) return undefined;
  const [bank, entry, field] = parts;
  if (!bank || !entry || !field) return undefined;
  return [bank, entry, field];
}

export function sampleBank(design: DesignDefinition, bank: string): SamplesDecl | undefined {
  return design.samples.get(bank);
}

export function sampleEntry(
  design: DesignDefinition,
  bank: string,
  entry: string,
): SampleEntryDecl | undefined {
  return sampleBank(design, bank)?.entries.find((e) => e.name === entry);
}

export function sampleField(
  design: DesignDefinition,
  bank: string,
  entry: string,
  field: string,
): SampleFieldDecl | undefined {
  return sampleEntry(design, bank, entry)?.fields.find((f) => f.name === field);
}

/** Resolve `Bank.entry.field` to the field decl, or a stable **PDL-E041** error. */
export function lookupSampleField(design: DesignDefinition, path: string): SampleFieldDecl {
  const split = splitSamplePath(path);
  if (!split) {
    throw new PdlError(
      "PDL-E041",
      `Invalid sample path \`${path}\` (expected Bank.entry.field)`,
      { path: design.entryPath },
    );
  }
  const [bank, entry, field] = split;
  if (!sampleBank(design, bank)) {
    throw new PdlError(
      "PDL-E041",
      `Unknown sample bank \`${bank}\` (path \`${path}\`)`,
      { path: design.entryPath },
    );
  }
  if (!sampleEntry(design, bank, entry)) {
    throw new PdlError(
      "PDL-E041",
      `Unknown sample entry \`${bank}.${entry}\` (path \`${path}\`)`,
      { path: design.entryPath },
    );
  }
  const f = sampleField(design, bank, entry, field);
  if (!f) {
    throw new PdlError("PDL-E041", `Unknown sample field \`${path}\``, { path: design.entryPath });
  }
  return f;
}

/** True when `path` is a three-segment path naming a known sample field. */
export function isKnownSamplePath(design: DesignDefinition, path: string): boolean {
  const split = splitSamplePath(path);
  if (!split) return false;
  const [bank, entry, field] = split;
  return sampleField(design, bank, entry, field) !== undefined;
}

File: crates/pdl-core/src/samples.rs
//! Typed sample banks (`samples Tracks { … }`) — lookup helpers for validate / evaluate / resolve.

use crate::ast::{SampleEntryDecl, SampleFieldDecl, SamplesDecl};
use crate::design::DesignDefinition;
use crate::error::PdlError;

/// Split `Bank.entry.field` into three segments. Longer/shorter dotted paths return `None`.
pub fn split_sample_path(path: &str) -> Option<(&str, &str, &str)> {
    let mut parts = path.split('.');
    let bank = parts.next()?;
    let entry = parts.next()?;
    let field = parts.next()?;
    if bank.is_empty() || entry.is_empty() || field.is_empty() {
        return None;
    }
    if parts.next().is_some() {
        return None;
    }
    Some((bank, entry, field))
}

pub fn sample_bank<'a>(design: &'a DesignDefinition, bank: &str) -> Option<&'a SamplesDecl> {
    design.samples.get(bank)
}

pub fn sample_entry<'a>(
    design: &'a DesignDefinition,
    bank: &str,
    entry: &str,
) -> Option<&'a SampleEntryDecl> {
    sample_bank(design, bank)?
        .entries
        .iter()
        .find(|e| e.name == entry)
}

pub fn sample_field<'a>(
    design: &'a DesignDefinition,
    bank: &str,
    entry: &str,
    field: &str,
) -> Option<&'a SampleFieldDecl> {
    sample_entry(design, bank, entry)?
        .fields
        .iter()
        .find(|f| f.name == field)
}

/// Resolve `Bank.entry.field` to the field decl, or a stable **PDL-E041** error.
pub fn lookup_sample_field<'a>(
    design: &'a DesignDefinition,
    path: &str,
) -> Result<&'a SampleFieldDecl, PdlError> {
    let Some((bank, entry, field)) = split_sample_path(path) else {
        return Err(PdlError::new(
            "PDL-E041",
            format!("Invalid sample path `{path}` (expected Bank.entry.field)"),
            Some(design.entry_path.clone()),
            None,
            None,
        ));
    };
    if sample_bank(design, bank).is_none() {
        return Err(PdlError::new(
            "PDL-E041",
            format!("Unknown sample bank `{bank}` (path `{path}`)"),
            Some(design.entry_path.clone()),
            None,
            None,
        ));
    }
    if sample_entry(design, bank, entry).is_none() {
        return Err(PdlError::new(
            "PDL-E041",
            format!("Unknown sample entry `{bank}.{entry}` (path `{path}`)"),
            Some(design.entry_path.clone()),
            None,
            None,
        ));
    }
    sample_field(design, bank, entry, field).ok_or_else(|| {
        PdlError::new(
            "PDL-E041",
            format!("Unknown sample field `{path}`"),
            Some(design.entry_path.clone()),
            None,
            None,
        )
    })
}

/// True when `path` is a three-segment path naming a known sample field.
pub fn is_known_sample_path(design: &DesignDefinition, path: &str) -> bool {
    match split_sample_path(path) {
        Some((bank, entry, field)) => sample_field(design, bank, entry, field).is_some(),
        None => false,
    }
}
